# 🧹 Étape 1 : Audit, Nettoyage et Pipeline ETL vers MySQL
**Projet :** Observatoire du Marché Automobile d'Occasion au Maroc (Avito)  
**Objectif :** Transformer le dataset brut issu du web scraping (**24 776 lignes, 32 colonnes**) en une table relationnelle propre (**8 097 annonces uniques, 26 colonnes**) prête pour l'entrepôt de données MySQL.

---
## 1. Chargement et Audit Initial des Données Brutes

In [36]:
import pandas as pd
import numpy as np
import re

In [37]:
df = pd.read_csv('avito_car_dataset_ALL.csv', encoding='latin1')

df.describe()

,Unnamed: 0,Année-Modèle,Puissance fiscale,Nombre de portes,Prix
count,24776.000000,24776.000000,24776.000000,20396.000000,2.477600e+04
mean,12387.500000,2011.844244,7.193938,4.905766,2.169588e+05
std,7152.359471,5.462925,1.574124,0.423789,6.749569e+06
min,0.000000,1990.000000,5.000000,3.000000,6.000000e+04
25%,6193.750000,2008.000000,6.000000,5.000000,7.500000e+04
50%,12387.500000,2013.000000,7.000000,5.000000,1.000000e+05
75%,18581.250000,2016.000000,8.000000,5.000000,1.400000e+05
max,24775.000000,2022.000000,14.000000,5.000000,6.120372e+08


In [38]:
df.head(10)

,Unnamed: 0,Lien,Ville,Secteur,Marque,Modèle,Année-Modèle,Kilométrage,Type de carburant,Puissance fiscale,...,Caméra de recul,Vitres électriques,ABS,ESP,Régulateur de vitesse,Limiteur de vitesse,CD/MP3/Bluetooth,Ordinateur de bord,Verrouillage centralisé à distance,Prix
0,0,https://www.avito.ma/fr/massira_2/voitures/FIA...,Temara,Massira 2,Fiat,Punto,2007,200 000 - 249 999,Diesel,5,...,False,False,True,False,False,False,True,False,False,60000
1,1,https://www.avito.ma/fr/temara/voitures/Dacia_...,Temara,NaN,Dacia,Dokker Van,2013,400 000 - 449 999,Diesel,6,...,False,False,False,False,False,False,False,False,False,70000
2,2,https://www.avito.ma/fr/casablanca/voitures/Da...,Casablanca,NaN,Dacia,Dokker,2014,160 000 - 169 999,Diesel,6,...,False,False,False,False,False,False,False,False,False,90000
3,3,https://www.avito.ma/fr/casablanca/voitures/to...,Casablanca,NaN,Volkswagen,Touareg,2005,0 - 4 999,Diesel,10,...,False,False,False,False,False,False,False,False,False,90000
4,4,https://www.avito.ma/fr/dakhla/voitures/Toyota...,Dakhla,NaN,Toyota,Prado,2007,200 000 - 249 999,Diesel,12,...,False,False,True,False,False,False,True,False,False,97000
5,5,https://www.avito.ma/fr/khouribga/voitures/Tig...,Khouribga,Toute la ville,Volkswagen,Tiguan,2014,180 000 - 189 999,Diesel,8,...,False,False,True,True,False,False,True,False,False,255000
6,6,https://www.avito.ma/fr/autre_secteur/voitures...,Meknès,Autre secteur,Peugeot,308,2009,200 000 - 249 999,Diesel,6,...,False,False,True,True,False,False,True,False,False,85000
7,7,https://www.avito.ma/fr/casablanca/voitures/Re...,Casablanca,NaN,Renault,Clio,2014,75 000 - 79 999,Diesel,6,...,False,False,True,True,False,False,True,False,False,93000
8,8,https://www.avito.ma/fr/mohammedia/voitures/Pe...,Mohammedia,NaN,Peugeot,208,2021,10 000 - 14 999,Diesel,6,...,False,False,True,True,False,False,True,False,False,164000
9,9,https://www.avito.ma/fr/fquih_ben_saleh/voitur...,Fquih Ben Saleh,NaN,Volkswagen,GOLF 7,2013,120 000 - 129 999,Diesel,8,...,False,False,True,True,False,False,True,False,False,198000


In [39]:
df.shape

(24776, 32)

## 2. Déduplication et Extraction de l'Identifiant Unique (Regex)
L'audit initial révèle la présence d'annonces republiées plusieurs fois. Nous supprimons la colonne d'index parasite `Unnamed: 0`, dédupliquons sur l'URL exacte (`Lien`), et extrayons l'identifiant numérique unique de chaque annonce via une expression régulière.

In [40]:
if 'Unnamed: 0' in df.columns:
    df= df.drop(columns=['Unnamed: 0'])

In [41]:
df = df.drop_duplicates(subset=['Lien']).reset_index(drop=True)
print(f"Lignes uniques après déduplication : {len(df)}")

Lignes uniques après déduplication : 8230


In [42]:
df['ID_Annonce'] = df['Lien'].str.extract(r'_(\d+)\.htm')[0]

## 3. Feature Engineering : Conversion du Kilométrage (Texte $\rightarrow$ Numérique)
Le kilométrage étant renseigné sous forme d'intervalles textuels (ex. `"200 000 - 249 999"`), nous extrayons les bornes inférieure (`Km_Min`), supérieure (`Km_Max`) et calculons le centre de classe (`Km_Moyen`).

In [43]:
def extraire_kilometrage(val):
    nombres = re.findall(r'\d+', str(val).replace(' ', ''))
    if len(nombres) == 2:
        km_min, km_max = float(nombres[0]), float(nombres[1])
        return pd.Series([km_min, km_max, (km_min + km_max) / 2.0])
    return pd.Series([np.nan, np.nan, np.nan])

df[['Km_Min', 'Km_Max', 'Km_Moyen']] = df['Kilométrage'].apply(extraire_kilometrage)
df.head()

,Lien,Ville,Secteur,Marque,Modèle,Année-Modèle,Kilométrage,Type de carburant,Puissance fiscale,Boite de vitesses,...,Régulateur de vitesse,Limiteur de vitesse,CD/MP3/Bluetooth,Ordinateur de bord,Verrouillage centralisé à distance,Prix,ID_Annonce,Km_Min,Km_Max,Km_Moyen
0,https://www.avito.ma/fr/massira_2/voitures/FIA...,Temara,Massira 2,Fiat,Punto,2007,200 000 - 249 999,Diesel,5,Manuelle,...,False,False,True,False,False,60000,49098076,200000.0,249999.0,224999.5
1,https://www.avito.ma/fr/temara/voitures/Dacia_...,Temara,NaN,Dacia,Dokker Van,2013,400 000 - 449 999,Diesel,6,Manuelle,...,False,False,False,False,False,70000,49098070,400000.0,449999.0,424999.5
2,https://www.avito.ma/fr/casablanca/voitures/Da...,Casablanca,NaN,Dacia,Dokker,2014,160 000 - 169 999,Diesel,6,Manuelle,...,False,False,False,False,False,90000,49098065,160000.0,169999.0,164999.5
3,https://www.avito.ma/fr/casablanca/voitures/to...,Casablanca,NaN,Volkswagen,Touareg,2005,0 - 4 999,Diesel,10,Automatique,...,False,False,False,False,False,90000,49098066,0.0,4999.0,2499.5
4,https://www.avito.ma/fr/dakhla/voitures/Toyota...,Dakhla,NaN,Toyota,Prado,2007,200 000 - 249 999,Diesel,12,Manuelle,...,False,False,True,False,False,97000,49098060,200000.0,249999.0,224999.5


# 4. Réduction de Dimension (Variance Nulle) & Score d'Équipement
Nous éliminons automatiquement les variables booléennes constantes (11 colonnes à variance nulle) et construisons un indicateur synthétique `Nb_Equipements` (score de 0 à 5 regroupant *Airbags, Climatisation, ABS, ESP, CD/MP3/Bluetooth*).

In [44]:
bool_cols = df.select_dtypes(include='bool').columns
colonnes_constantes = [col for col in bool_cols if df[col].nunique() == 1]
print(f"Colonnes supprimées (variance nulle) : {len(colonnes_constantes)} colonnes")
df = df.drop(columns=colonnes_constantes)

Colonnes supprimées (variance nulle) : 11 colonnes


In [45]:
equipements_actifs = ['Airbags', 'Climatisation', 'ABS', 'ESP', 'CD/MP3/Bluetooth']
df['Nb_Equipements'] = df[equipements_actifs].sum(axis=1)

## 5. Imputation des Valeurs Manquantes et Calcul de l'Âge
- Standardisation des valeurs manquantes déguisées (`"--"`) en `"Non spécifié"`.
- Imputation du `Nombre de portes` par la valeur modale spécifique au couple `(Marque, Modèle)`.
- Calcul de `Age_Vehicule` par rapport à l'année de référence maximale du parc observé (2022).

In [46]:
df['Boite de vitesses'] = df['Boite de vitesses'].replace('--', 'Non spécifié')


cols_cat_na = ['Secteur', 'Origine', 'Première main', 'État']
for col in cols_cat_na:
    df[col] = df[col].fillna('Non spécifié').astype(str).str.strip()

df['Nombre de portes'] = df.groupby(['Marque', 'Modèle'])['Nombre de portes'].transform(
    lambda x: x.fillna(x.mode()[0] if not x.mode().empty else 5.0)
).astype(int)

annee_reference = int(df['Année-Modèle'].max())
df['Age_Vehicule'] = annee_reference - df['Année-Modèle']
df.head()

,Lien,Ville,Secteur,Marque,Modèle,Année-Modèle,Kilométrage,Type de carburant,Puissance fiscale,Boite de vitesses,...,ABS,ESP,CD/MP3/Bluetooth,Prix,ID_Annonce,Km_Min,Km_Max,Km_Moyen,Nb_Equipements,Age_Vehicule
0,https://www.avito.ma/fr/massira_2/voitures/FIA...,Temara,Massira 2,Fiat,Punto,2007,200 000 - 249 999,Diesel,5,Manuelle,...,True,False,True,60000,49098076,200000.0,249999.0,224999.5,3,15
1,https://www.avito.ma/fr/temara/voitures/Dacia_...,Temara,Non spécifié,Dacia,Dokker Van,2013,400 000 - 449 999,Diesel,6,Manuelle,...,False,False,False,70000,49098070,400000.0,449999.0,424999.5,0,9
2,https://www.avito.ma/fr/casablanca/voitures/Da...,Casablanca,Non spécifié,Dacia,Dokker,2014,160 000 - 169 999,Diesel,6,Manuelle,...,False,False,False,90000,49098065,160000.0,169999.0,164999.5,0,8
3,https://www.avito.ma/fr/casablanca/voitures/to...,Casablanca,Non spécifié,Volkswagen,Touareg,2005,0 - 4 999,Diesel,10,Automatique,...,False,False,False,90000,49098066,0.0,4999.0,2499.5,0,17
4,https://www.avito.ma/fr/dakhla/voitures/Toyota...,Dakhla,Non spécifié,Toyota,Prado,2007,200 000 - 249 999,Diesel,12,Manuelle,...,True,False,True,97000,49098060,200000.0,249999.0,224999.5,3,15


## 6. Filtrage Économétrique des Valeurs Aberrantes (IQR par Marque)
Application d'un double filtre pour éliminer les erreurs de saisie manifestes (ex. prix saisis en centimes ou à 1 DH) :
1. **Cadrage global du segment d'occasion standard :** $\text{Prix} \in [60\,000\text{ MAD} ; 600\,000\text{ MAD}]$.
2. **Filtre Inter-Quartile (IQR $\times 3$) conditionnel par Marque :** Élimine les valeurs extrêmes tout en respectant la distribution de prix propre à chaque constructeur (ex. *Dacia* vs *Mercedes-Benz*).

In [47]:
df = df[(df['Prix'] >= 60000) & (df['Prix'] <= 600000)]

q1 = df.groupby('Marque')['Prix'].transform(lambda x: x.quantile(0.25))
q3 = df.groupby('Marque')['Prix'].transform(lambda x: x.quantile(0.75))
seuil_haut = q3 + 3.0 * (q3 - q1)

df = df[df['Prix'] <= seuil_haut].reset_index(drop=True)
print(f"Dimensions après filtrage : {df.shape[0]} lignes et {df.shape[1]} colonnes")

Dimensions après filtrage : 8097 lignes et 26 colonnes


## 7. Export CSV & Ingestion Automatisée vers MySQL (`staging_avito`)
Normalisation des noms de colonnes (sans accents ni espaces) et chargement direct dans la couche de *Staging* de l'entrepôt MySQL `avito_bi_dw` via `SQLAlchemy`.

In [48]:
import urllib.parse
from sqlalchemy import create_engine, text

df_sql = df.copy()
df_sql.to_csv('avito_car_clean_final.csv', index=False, encoding='utf-8')
print(f"Fichier CSV propre sauvegardé : {df_sql.shape[0]} lignes et {df_sql.shape[1]} colonnes.")

rename_map = {
    'Lien': 'Lien',
    'Ville': 'Ville',
    'Secteur': 'Secteur',
    'Marque': 'Marque',
    'Modèle': 'Modele',
    'Année-Modèle': 'Annee_Modele',
    'Kilométrage': 'Kilometrage',
    'Type de carburant': 'Type_Carburant',
    'Puissance fiscale': 'Puissance_Fiscale',
    'Boite de vitesses': 'Boite_Vitesses',
    'Nombre de portes': 'Nombre_Portes',
    'Origine': 'Origine',
    'Première main': 'Premiere_Main',
    'État': 'Etat',
    'Airbags': 'Airbags',
    'Climatisation': 'Climatisation',
    'ABS': 'ABS',
    'ESP': 'ESP',
    'CD/MP3/Bluetooth': 'CD_MP3_Bluetooth',
    'Prix': 'Prix',
    'ID_Annonce': 'ID_Annonce',
    'Km_Min': 'Km_Min',
    'Km_Max': 'Km_Max',
    'Km_Moyen': 'Km_Moyen',
    'Nb_Equipements': 'Nb_Equipements',
    'Age_Vehicule': 'Age_Vehicule'
}
df_sql = df_sql.rename(columns=rename_map)

bool_cols = ['Airbags', 'Climatisation', 'ABS', 'ESP', 'CD_MP3_Bluetooth']
df_sql[bool_cols] = df_sql[bool_cols].astype(int)

password = urllib.parse.quote_plus("SQL_PASSWORD")
engine = create_engine(f"mysql+pymysql://root:{password}@localhost:3306/avito_bi_dw?charset=utf8mb4")

with engine.begin() as conn:
    conn.execute(text("DROP TABLE IF EXISTS staging_avito;"))
    df_sql.to_sql(name='staging_avito', con=conn, if_exists='replace', index=False)

print(f"Succès total ! {len(df_sql)} lignes et {df_sql.shape[1]} colonnes insérées dans MySQL (staging_avito) !")

Fichier CSV propre sauvegardé : 8097 lignes et 26 colonnes.
Succès total ! 8097 lignes et 26 colonnes insérées dans MySQL (staging_avito) !
